### PR TITLE
[Help] Add simulated slanted font support to the editor help.

### PR DESCRIPTION
**Pull request by bruvzg**
Screenshot 2022-03-11 at 16 24 03

Fixes (partially, only missing italic) #27456

### CODE REVIEW

**Comment by akien-mga** [comment]
You can likely remove the color override, I think it was used only to differentiate with normal text since we lacked italics. It also seems to cause a bug as the part after `radians` in your example also seems to use the italics color (unless it's an optical illusion).

File: editor/editor_fonts.cpp [hunk 1]
```diff
@@ -263,6 +311,31 @@ void editor_register_fonts(Ref<Theme> p_theme) {
 	Ref<FontData> FontThai = load_cached_internal_font(_font_NotoSansThaiUI_Regular, _font_NotoSansThaiUI_Regular_size, font_hinting, font_antialiased, true, font_subpixel_positioning);
 	Ref<FontData> FontThaiBold = load_cached_internal_font(_font_NotoSansThaiUI_Bold, _font_NotoSansThaiUI_Bold_size, font_hinting, font_antialiased, true, font_subpixel_positioning);
 
+	Ref<FontData> DefaultFontSlanted = DefaultFont->duplicate();
+	DefaultFontSlanted->set_transform(Transform2D(1.0, 0.3, 0.0, 1.0, 0.0, 0.0));
```
Comment (akien-mga) [inline editor/editor_fonts.cpp:315]: I guess you could factor this out in `const Transform2D slanted_transform = Transform2D(1.0, 0.3, 0.0, 1.0, 0.0, 0.0);`

**Comment by bruvzg** [comment]
Removed the color, I have left it for non-Latin scripts, but it seems at least some of them use slanted fonts as well, so I have added slanted variant to all languages.

Screenshot 2022-03-11 at 17 18 21

File: editor/editor_fonts.cpp (hunk 1 above)
Comment (bruvzg) [reply editor/editor_fonts.cpp]: Done

**Comment by akien-mga** [comment]
> Removed the color, I have left it for non-Latin scripts, but it seems at least some of them use slanted fonts as well, so I have added slanted variant to all languages.

Ah that's a good point, we might need to review this for each language.

AFAIK CJK doesn't use italics (CC @timothyqiu) so for those we might want to use something else for emphasis.

I don't know if say Arabic uses italics for emphasis but if it does, I'd imagine that it would be slanted to the left.

**Comment by timothyqiu** [comment]
> AFAIK CJK doesn't use italics (CC @timothyqiu) so for those we might want to use something else for emphasis.

Yeah, italics for Chinese characters are usually considered bad practice, it hurts legibility. It's common to use a completely different font for emphasis, but I think bold is also acceptable.

**Comment by bruvzg** [comment]
> AFAIK CJK doesn't use italics (CC @timothyqiu) so for those we might want to use something else for emphasis.

I have not applied it to CJK (we also do not have bold CJK fonts). A quick search found that slanted/italic Arabic, Bengali, Devanagari, Georgian, Hebrew, Tamil and Thai fonts exist. No idea about the rest.

**Comment by akien-mga** [comment]
> A quick search found that slanted/italic Arabic, Bengali, Devanagari, Georgian, Hebrew, Tamil and Thai fonts exist.

For Arabic, I think it might be more expected if the font was slanted to the left. I've seen examples with it slanted both towards the top left or the bottom left.

See Adobe Arabic (and its "Iranic" form):
image

Lots of useful insights in https://typedrawers.com/discussion/2147/where-is-arabic-italic-originating-from and the TL;DR seems to be - maybe it's not as meaningful for Arabic readers as we might think, and maybe better left out/just with a different color?

CC @khaledhosny if you want to provide some insights on do's and don'ts for Arabic scripts emphasis. For context, we're using simulated slanting instead of proper italics/cursive typeface mostly for binary size reasons, and because italics/emphasized text is fairly uncommon in our in-editor documentation (which this PR affects).

**Comment by bruvzg** [comment]
> For Arabic, I think it might be more expected if the font was slanted to the left. I've seen examples with it slanted both towards the top left or the bottom left.



Seems like it's not always the case, but it should be easy to do with simulated italic (just change slant to negative).

**Comment by bruvzg** [comment]
For the reference, screenshots of the editor Arabic font:

| Original Font | Slanted 0.4 right (this PR) | Slanted 0.4 left |
|---|---|---|
| Screenshot 2022-03-18 at 19 20 59 | Screenshot 2022-03-18 at 19 21 31 | Screenshot 2022-03-18 at 19 21 39 |

**Comment by khaledhosny** [comment]
Although slanted Arabic exists, it was not traditionally used as a contrastive style in the way italic come to be used in Latin. Slanting is poor for giving contrast in Arabic since many letters don’t have vertical stems where the slanting would make a noticeable difference, and most Arabic fonts don’t include “italic” versions probably for this reason (and when they do, it is usually auto-slanted anyway).

Slanting to the left is more natural for Arabic, but most applications don’t bother and slant Arabic to the right. Slanting to the left need to be carefully done to not end up with glyphs in the same run of text being slanted to different direction (e.g. number in an Arabic script run should be left slanted like the Arabic even though they get LTR direction).

My recommendation for content creators is to avoid slanted Arabic, it is a bad practice, but for software is to apply synthetic slanting if requested.

**Comment by khaledhosny** [comment]
For synthetic slanting, you should also use `hb_font_set_synthetic_slant()` to apply the slant to glyph positions (see examples in https://github.com/harfbuzz/harfbuzz/pull/3338 for the difference it makes).

**Comment by bruvzg** [comment]
Added `hb_font_set_synthetic_slant()` call, it seems to improve diacritic placement for the slanted text.

Screenshot 2022-03-22 at 14 14 48

For now, replaced all non-Latin slanted fonts with bold fonts.

**Review by akien-mga — APPROVED** [review]

File: editor/editor_fonts.cpp [hunk 2]
```diff
@@ -269,6 +317,11 @@ void editor_register_fonts(Ref<Theme> p_theme) {
 	Ref<FontData> FontThai = load_cached_internal_font(_font_NotoSansThaiUI_Regular, _font_NotoSansThaiUI_Regular_size, font_hinting, font_antialiased, true, font_subpixel_positioning);
 	Ref<FontData> FontThaiBold = load_cached_internal_font(_font_NotoSansThaiUI_Bold, _font_NotoSansThaiUI_Bold_size, font_hinting, font_antialiased, true, font_subpixel_positioning);
 
+	const Transform2D slanted_transform = Transform2D(1.0, 0.3, 0.0, 1.0, 0.0, 0.0);
+
+	Ref<FontData> DefaultFontSlanted = DefaultFont->duplicate();
+	DefaultFontSlanted->set_transform(slanted_transform);
```
Comment (akien-mga) [inline editor/editor_fonts.cpp:323]: Now it's only used once so I guess the constant is no longer needed.

**Comment by akien-mga** [comment]
Thanks!

And thanks @khaledhosny for your valuable input!

**Pull request merged**